src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/ViewDataDictionary.cs

```diff
@@ -395,37 +455,54 @@ public ViewDataInfo GetViewDataInfo(string expression)
             return ViewDataEvaluator.Eval(this, expression);
         }
 
-        // This method will execute before the derived type's instance constructor executes. Derived types must
-        // be aware of this and should plan accordingly. For example, the logic in SetModel() should be simple
-        // enough so as not to depend on the "this" pointer referencing a fully constructed object.
+        /// <summary>
+        /// Set <see cref="ModelExplorer"/> to ensure <see cref="Model"/> and <see cref="ModelExplorer.Model"/>
+        /// reflect the new <paramref name="value"/>.
+        /// </summary>
+        /// <param name="value">New <see cref="Model"/> value.</param>
         protected virtual void SetModel(object value)
         {
-            EnsureCompatible(value);
-
-            // Reset or override ModelMetadata based on runtime value type. Fall back to declared type if value is
-            // null. When called from a constructor, current ModelExplorer may already be set to preserve
-            // customizations made in parent scope. But ModelExplorer is never null after instance is initialized.
-            var modelType = GetModelType(value);
-            Type metadataModelType = null;
-            if (ModelExplorer != null)
+            // Update ModelExplorer to reflect the new value. When possible, preserve ModelMetadata to avoid losing
+            // property information.
+            var modelType = value?.GetType();
+            if (ModelMetadata.MetadataKind == ModelMetadataKind.Type &&
```

Comment:
Note these conditions are now in the same order and identical to those in the copy-from-`source` constructor.